File: 2022/aoc_python/day_6.py
from lib import io

print("Advent of Code 2022 Day 6")
print("-------------------------")
signal_input = io.file_to_string("input/day-6-input.txt")

def find_unique_substring_starting_index(str, substring_length):
    starting_index = substring_length
    while starting_index < len(str):
        if len(set(str[starting_index - substring_length : starting_index])) == substring_length:
            # Found start-of-substring marker (since sets do not hold duplicates).
            break
        else:
            starting_index += 1

    return starting_index

# Sweep a window of size 4 across signal_input, looking for a substring of 4 unique characters.
start_of_packet_index = find_unique_substring_starting_index(str=signal_input, substring_length=4)

print("PART 1")
print("======")
print(f"First marker index: {start_of_packet_index}")
print("======")
# Attempt 1: 1658.

File: 2022/aoc_scala/Day6.scala
import scala.collection.Set
import scala.io.Source
import scala.util.control.Breaks._

def findUniqueSubstringStartingIndex(str: String, substringLength: Int): Int = 
    var startingIndex = substringLength
    breakable {
        while (startingIndex < str.length()) {
            if (str.substring(startingIndex - substringLength, startingIndex).toSet.knownSize == substringLength) {
                // Found start-of-substring marker (since sets do not hold duplicates).
                break
            } else {
                startingIndex += 1
            }
        }
    }

    startingIndex

@main def daySix() =
    println("Advent of Code 2022 Day 6")
    println("-------------------------")
    val signalInput = Source.fromFile("input/day-6-input.txt").mkString
    // println(s"signalInput: $signalInput")

    // Sweep a window of size 4 across signal_input, looking for a substring of 4 unique characters.
    val startOfPacketIndex = findUniqueSubstringStartingIndex(signalInput, 4)

    println("PART 1")
    println("======")
    println(s"First marker index: $startOfPacketIndex")
    println("======")


    // Sweep a window of size 14 across signal_input, looking for a substring of 14 unique characters.
    val startOfMessageIndex = findUniqueSubstringStartingIndex(signalInput, 14)

    println("PART 2")
    println("======")
    println(s"First marker index: $startOfMessageIndex")
